src/RawPOSIX/src/interface/signal.rs

```diff
@@ -0,0 +1,243 @@
+// returns an optional tuple where the first element is the signal number
+// the second element is the signal handler
+// and the third element is the signal mask restore callback function
+pub fn lind_get_first_signal(cageid: u64) -> Option<(i32, u32, Box<dyn Fn(u64)>)> {
+    let cage = cagetable_getref(cageid);
+    let mut pending_signals = cage.pending_signals.write();
+    let sigset = cage.sigset.load(RustAtomicOrdering::Relaxed);
+
+    // we iterate through signal and retrieve the first unblocked signals in the pending list
+    if let Some(index) = pending_signals.iter().position(
+        |&signo| (sigset & convert_signal_mask(signo)) == 0 // check if signal is blocked
+    ) {
+        // retrieve the signal number
+        let signo = pending_signals.remove(index);
+        // retrieve the corresponding signal handler
+        match cage.signalhandler.get_mut(&signo) {
+            Some(mut sigaction) => {
+                // if sigprocmask is called during the execution of the signal handler
+                // the signal mask will not be perseved once handler is finished
+
+                // by default, we block the same signal during its execution
+                let mut mask_self = convert_signal_mask(signo);
+                let signal_handler = sigaction.sa_handler;
+                // if SA_RESETHAND is set, we reset the signal handler to default for this signal
+                if sigaction.sa_flags as u32 & SA_RESETHAND > 0 {
+                    sigaction.sa_handler = SIG_DFL as u32;
+                }
+
+                // if SA_NODEFER is set, we allow the same signal to interrupt itself
+                if sigaction.sa_flags as u32 & SA_NODEFER > 0 {
+                    mask_self = 0;
+                }
+                // temporily update the signal mask
+                cage.sigset.fetch_or(sigaction.sa_mask | mask_self, RustAtomicOrdering::Relaxed);
+                
+                // restorer is called when the signal handler finishes. It should restore the signal mask
+                let restorer = Box::new(move |cageid| {
+                    let cage = cagetable_getref(cageid);
+                    cage.sigset.store(sigset, RustAtomicOrdering::Relaxed);
+                });
+                Some((signo, signal_handler, restorer))
+            }
+            None => {
+                // retrieve the signal handler
+                let signal_handler = signal_get_handler(cageid, signo);
+                let restorer = Box::new(move |cageid| {
+                    let cage = cagetable_getref(cageid);
+                    cage.sigset.store(sigset, RustAtomicOrdering::Relaxed);
+                });
+                Some((signo, signal_handler, restorer))
+            }
+        }
+    } else {
+        // if there is no pending unblocked signal, we return None
+        None
+    }
+}
+
+// check if there is any pending unblocked signals
+pub fn lind_check_no_pending_signal(cageid: u64) -> bool {
```

Comment:
confusing function name, comment needs a bit more of whats happening here and retval